### DISCUSSION TITLE
Possibility to build with library evolution enabled

### DISCUSSION THREAD

**Discussion by flockoffiles** [pointfreeco/swift-composable-architecture#1442] [Beta]
In some situations it can be useful to compile Composable Architecture into a binary .xcframework and import it like that.

The current version of Composable Architecture can be built with -library-evolution option enabled, but the upcoming "protocol" branch no longer builds.

For previous discussions related to xcframeworks, please see: https://github.com/pointfreeco/swift-composable-architecture/discussions/890

Also see this comment discussion: https://github.com/pointfreeco/swift-composable-architecture/pull/1283#discussion_r986866267

**Comment by mbrandonw** [comment]
And just to confirm, is building with library evolution a requirement in order to build an xcframework? My impression is that it is not, but that if you do you may get a warning about binary compatibility not being guaranteed.

**Reply by flockoffiles** [reply]
I have not tried it yet (building an .xcframework without library evolution enabled). One quite obvious immediate limitation is that such a framework will become dependent on a specific compiler/xcode version, so the binary built with, say Xcode 14.0 may not correctly link with a module being built with XCode 14.1, etc.

**Reply by mbrandonw** [reply]
That's true, but also building with library evolution when TCA doesn't promise library evolution is bound to have its own problems too, right? That means you could swap in a new TCA without re-building the thing that depends on TCA, which will lead to crashes at runtime.

**Reply by flockoffiles** [reply]
In the most generic case, you are totally right - you can't simply swap for a new TCA without rebuilding. That's the limitation that in our particular case is not very important. But having to depend on a specific compiler/Xcode version for the same version of TCA is quite a bummer...
But once again, if making the framework compile with library evolution enabled presents a considerable effort on your side, you may indeed, choose to not support it at all...
Perhaps, Apple will eventually fix the compiler, so that the same file can compile correctly in either mode without us having to invent workarounds. (That's at least an expectation for most people out there).

**Reply by mbrandonw** [reply]
I'm not sure it's a considerable amount of work yet. We would definitely need to add another CI step to build the library with evolution on so that we can catch regressions. I'm just wanting to make sure I understand the full problem.

Do you know of an easy way to build for library evolution directly in Xcode? I only know how to do it from the command line, which is a bit of a pain.

**Reply by flockoffiles** [reply]
For an ordinary framework target, you set "Build Libraries for Distribution" to Yes under "Build Settings" in XCode.

For Swift Package, you probably need to specify the compiler option explicitly, just like you do when you enable concurrency warnings, etc.
(Have not done it myself, but here seems to be an example: https://github.com/apple/swift-package-manager/blob/main/Package.swift)
(Search for -enable-library-evolution)

**Reply by mbrandonw** [reply]
Took a stab at it [here](https://github.com/pointfreeco/swift-composable-architecture/pull/1443/), but Stephen is out for a few days so may not merge it right away. But in the meantime can you point your project to that branch and let me know if it works?

**Reply by flockoffiles** [reply]
Yes. Trying...

**Reply by flockoffiles** [reply]
It builds fine now! You are SUPER awesome! :-)
